content/200-concepts/100-components/02-prisma-client/260-client-extensions/index.mdx

```diff
@@ -0,0 +1,170 @@
+
+- If you use two or more middlewares at the same time, then they all apply globally to the same client
+- Extensions are isolated: the functionality in each one is separate, unless you deliberately combine them
+- All extended clients share the same connection pool, but their functionality is separate
+- In the RLS example given above, with extensions each user can be in an entirely separate client. With middlewares, all users are active in the same client.
+
+## Share Prisma Client extensions
+
+You can share Prisma Client extensions with other users, either as packages or as code snippets. To create an extension that is more likely to work for other users, use the `$all` prefix, for example [`$allModels`](/concepts/components/prisma-client/client-extensions/client-extensions-model#add-a-custom-method-to-all-models-in-your-schema). This extension applies generally, not just to your specific schema.
+
+For example, a generic extension might take the following form:
+
+```ts
+const xprisma = new prisma.$extends({
+  model: {
+    $allModels: {
+      getClass<T extends object>(this: T): new () => T {
+        return class {
+          /** Generic Implementation */
+        } as any
+      },
+    },
+  },
+})
+
+class UserService extends xprisma.user.getClass() {
+  method() {
+    const user = this.findFirst({}) // fully type-safe
+  }
+}
+```
+
+### Package an extension
+
+To package an extension, use `Prisma.createExtension`. This is very similar to `prisma.$extends`, but it declares an extension.
+For example:
+
+```ts
+
+import { Prisma } from '@prisma/client'
+
+export default Prisma.createExtension({
+  model: {
+    $allModels: {
+      // new method
+      findOrCreate(...) { } // code for the new method goes inside the brackets
+    }
+  }
+})
+```
+
+You could then share the above code on `npm` or as a code snippet.
+
+### Install a packaged extension
+
+In your project, you can install any Prisma Client extension that another user has published to `npm`. To do so, add the following line to your application code:
+
+```ts
+npm install prisma-extension-<name>
+```
+
+For example, if an available extension is called `find-or-create`, you could install it as follows:
+
+```ts
+npm install prisma-extension-find-or-create
+```
+
+To import the `find-or-create` extension from the example above, and load it into your client instance, you could use the following code:
+
+```ts
+import findOrCreate from 'prisma-extension-find-or-create'
+
+const xprisma = prisma.$extends(findOrCreate)
+const user = await xprisma.user.findOrCreate({ ... })
+
+xprisma.user.findOrCreate()
+```
+
+<Admonition type = "info">
+
+When you call a method in your extension, you must use the correct instance in your code. In the above example, `xprisma.user.findOrCreate` works, but `prisma.user.findOrCreate` does not.
+
+</Admonition>
```

Comment:
Is this specific for extensions installed via a package? Otherwise I think this should come earlier in a more general section about how extensions work.

Reply:
I included this information after every relevant example, because a user might well jump straight to the example. I want to make the examples as self-contained as possible. However, I agree that this information should also go into the general material on extensions.

Reply:
I'll also reword this to make it clearer that the text refers to the nuts and bolts of this example - it's not wider point about installing extensions.